backend/in_memory_store.py

```diff
@@ -0,0 +1,233 @@
+  '/' or '', obj is returned.
+  """
+  if path.endswith('/'):
+    path = path[:-1]
+  path_parts = path.split('/')[1:]
+  for path_part in path_parts:
+    if not path_part in obj:
+      if create_missing:
+        obj[path_part] = {}
+      else:
+        return None
+    obj = obj[path_part]
+  return obj
+
+def join_paths(path, suffix):
+  """Joins a path and a suffix into a single path.
+  Resolves leading and trailing '/'. The suffix can be null.
+
+  Examples:
+    join_paths('abc/def/', '/hij') => 'abc/def/hij'
+    join_paths('a', 'b/c/d') => 'a/b/c/d'
+    join_paths('abc/def', None) => 'abc/def/'
+  """
+  if not path.endswith('/'):
+    path = path + '/'
+  if suffix.startswith('/'):
+    suffix = suffix[1:]
+  if suffix is None:
+    suffix = ''
+  return '%s%s' % (path, suffix)
+
+def drop_last(path):
+  """Returns a path with the last "part" dropped.
+
+  Examples:
+    drop_last('abc/def/ghi') => 'abc/def'
+    drop_last('abc') => ''
+    drop_last('') => ''
+  """
+  if '/' not in path:
+    return ''
+  return '/'.join(path.split('/')[:-1])
+
+def last(path):
+  """Returns a last "part" of a path.
+
+  Examples:
+    last('abc/def/ghi') => 'ghi'
+    last('abc') => 'abc'
+    last('') => ''
+  """
+  if '/' not in path:
+    return path
+  return path.split('/')[-1]
+
+class InMemoryStore:
+  """An in memory version of the data in firebase. Mutations applied to the
+  store will also apply to the remote version of the data.
+  Getting data from the store will return the same results as getting data
+  from the remote firebase store.
+
+  CAVEAT:
+  Getting data during a transactional mutation (with several mutations)
+  locally will return data as if the current mutations in the transaction have
+  already been applied (because they have been). Getting data remotely will
+  return data as if none of the current mutations have been applied.
+  """
+  def __init__(self):
+    self.instance = None
+    self.firebase = None
+
+  def maybe_load(self, firebase):
+    """Load the firebase model from the remote source if a local copy
+    doesn't exist.
+    """
+    if self.instance is None and self.firebase is None:
+      print '*************** LOADING INSTANCE FROM FIREBASE *******************'
+      self.instance = firebase.get('/', None)
+      self.firebase = firebase
+
+  def get(self, path, id, params=None, local_instance=False):
+    """Get data from the model. Getting data from the local instance and the
+    remove instance is the same with some exceptions:
+    - Remote fetches don't have mutations from an unclosed transaction,
+      local fetches do.
+    - params don't do anything in the local instance
+
+    Arguments:
+      path: The path to get
+      id: The id of the value to get at the path
+      params: Filtering and sorting params
+      local_instance: If true, gets the data from the local copy,
+          if false, gets the data from the remote copy.
+
+    TODO(yuhao93): Make params work locally
```

Comment:
(just to have it written down somewhere)

feel free to not support orderby and filter, lets let the caller do that